### ISSUE TITLE
Pass %RAD_CONFIG and %RAD_REPLY to python modules in rlm_python

### ISSUE THREAD

**Issue by alejandro-perez** [FreeRADIUS/freeradius-server#1464]
Hi,

in rlm_pyhon, modules only have access to RAD_REQUEST. This is generally ok, but sometimes you need access to RAD_REPLY or even RAD_CONFIG (indeed, that's the modules's interface, so why would not do it?)

I have seen that rlm_perl does provide them all to sub-modules.

Thanks

**Comment by herwinw** [comment]
Don't forget RAD_STATE, RAD_REQUEST_PROXY and RAD_REQUEST_PROXY_REPLY that are available in Perl as well. While the architecture of `rlm_perl` makes it pretty easy to add another list to the input (global variables), `rlm_python` works with parameters that are passed to a function. This means we'd have to change the prototype of the function (or use the `*args, **kwargs`-feature).

I'm working on a proposal to change the interface of all the interfaces of the embedded languages, which includes this feature. I hope to have it finished this weekend.

**Comment by alejandro-perez** [comment]
Or you can change the "p" parameter from tuple list to map of tuple lists.
p["REQUEST"] = what we already have
p["REPLY"] = ...
p["RAD_STATE"] = ....

**Comment by alejandro-perez** [comment]
Hi, is this moving forward? 
In case it is not...is there any effective way by which I can know, from within a python module, whether the request is being proxied or not?
Thanks

**Comment by alandekok** [comment]
When you open an issue, there is no guarantee that someone will jump on it, and immediately issue a fix.

This is open source software.  You're welcome to submit a patch.

**Comment by alejandro-perez** [comment]
Hi Alan,

sorry if my comment sounded disrespectful, that was way far from being my intention. I was asking just because he mentioned that he might do it and I wanted to know if he was on the process. I know how opensource works, and I never mean to make any exigency. Again, sorry if it seemed to be the case. It wasn't.

Regards

**Comment by gyx1000** [comment]
I would like try to fix this issue. The problem is that we can't simply pass new arguments because we will introduce a break if python functions (authorize, authenticate, ...) doesn't assume it.

@arr2036 @alandekok What do you think about inspect the python functions before calling it and "inject" params according to the python functions signatures ?

**Comment by arr2036** [comment]
Could be interesting.  Do you know if it's possible to do lazy martialling and unmartialling in python? i.e. intercept hash operations (get, set, remove) and call out to C functions?

**Comment by gyx1000** [comment]
In python it's possible to intercept easily this kind of operation `__getattr__`, `__setattr__`, etc.... but I have no experience with mapping this calls directly to a C functions. I don't know if it's really possible without creating a custom `PyTypeObject`. Did I understand your question ?

What is your idea ?

**Comment by alejandro-perez** [comment]
Hi,

this issue has become relevant to me again. Is this something you plan to implement for v4 or are there plans to include it in v3?

**Comment by alandekok** [comment]
we probably can't add it in v3, unless we can do it without changing / breaking anything.

v4 is a better possibility.

But... it will likely be the end of September before we can even begin to look at this.  We're booked with v4 rearchitecture until then.

**Comment by alejandro-perez** [comment]
> we probably can't add it in v3, unless we can do it without changing / breaking anything.

I'll try to look at it in September and see if I can find a way. I think v3 lifecycle is not ended yet, so it will be worth it for us. If I can't, I may try to implement what I want in Perl, but I'm not a huge fan :)

> v4 is a better possibility.

Probably yes, given you are more actively involved in it now.

> But... it will likely be the end of September before we can even begin to look at this. We're booked with v4 rearchitecture until then.

Sure I completely understand. Thanks

**Comment by herwinw** [comment]
In `rlm_mruby` I've created a request object that has the (read-only) lists as attributes, so code could like like this:

```
def authorize(request)
    radlog(L_DBG, "Username: #{request.request['User-Name']}")
    radlog(L_DBG, "Something from session-state: #{request.session_state['Tmp-String-0']}")
```

(With `#{ ... }` being the Ruby syntax for embedding evaluation inside a string). This way the method signature will remain the same, even if lists are added or removed. Things like lazy conversions from C to internal Python objects could easily be added afterwards.

I might try to create something like this for v4 in the next month, but I'm a bit short on free time lately.

**Comment by alejandro-perez** [comment]
The solution makes sense to me, but as you say, it would not work for v3 as existing Python modules would not work. If a module is expecting "request" to be a list of VPs, now it is an object that contains multiple VPs, and the expected values are in request.request.

Not sure how difficult would it be to pass in a subclass of "tuple" which contains the other lists as attributes, but the request list still remains as the object itself. This way legacy modules would still work whereas if you know that request.reply will be there (you can even ask with hasattr()), you can use it.

**Comment by alejandro-perez** [comment]
Another option would be to add a configuration item, similar to the pythonpath, which indicates whether this module is using the "new" API or the "old" API (defaults to the latter).

Then rlm_python would know how to build the argument list for the callbacks.

**Comment by alejandro-perez** [comment]
I'm already working on a prototype of this latter option, which I think its the best alternative. If the option is not provided, a single Tuple with the list of attributes from request->packet->vps is provided. If the option is enabled, then a 6-tuple is provided (Request, Reply, Config, State, Proxy-Request, Proxy-Reply), where None is used for the non present lists.

Do you think a solution such as this could be introduced in v3.0.x? For v4.0.x you could remove the option and always go for the 6-tuple, which can be easily expand to 7-tuple in the future, etc.

**Comment by gyx1000** [comment]
Hello,

Maybe you can look in my closed PR #1609. I did this patch some time ago to close this issue.

Have a nice day

**Comment by alejandro-perez** [comment]
> Hello,
> 
> Maybe you can look in my closed PR #1609. I did this patch some time ago to close this issue.
> 
> Have a nice day

Indeed I did :), thanks. Why was it closed without being accepted? I couldn't find any discussion. Did you have a conversation off-line?

In any case, I saw your approach was slightly different, as you either call to method(p) or method(p1, p2, p3, p4...) by introspecting the Python module and see what the module expects. I don't disagree with it either, but it felt more natural to me to keep the signature and add this explicit module configuration parameter. In any case, I'm interested on having access to this information, so both solutions would work for me.

**Comment by gyx1000** [comment]
I don't know why it was closed.

Effectivley, your approach is more natural. ;)
Thanks for your job.

**Comment by alejandro-perez** [comment]
Well, I haven't done it yet, but I'm watching to your code and getting (i.e. copying) a couple of good ideas.

**Comment by alejandro-perez** [comment]
Umm, I'm rethinking this and not sure that my approach is better than yours...
I'm interested on hearing from the project managers here.

**Comment by alandekok** [comment]
I don't know much about python or it's API, so I can't offer help there.

The main thing for v3 is backwards compatibility.  If there's a flag saying "use new functionality", and the flag is off by default, that's fine.

**Comment by alejandro-perez** [comment]
Ok, I'll prepare a PR with such a functionality. I hope it will be ready today or tomorrow.

**Comment by alejandro-perez** [comment]
Done. You can find it in #2031

**Comment by alejandro-perez** [comment]
At least, this is closed for v3.0.x. For v4.0.x either you can use the code for v3.0.x or you can do it differently.
